tests/Owin.Security.OpenIdConnect.Server.Tests/OpenIdConnectMessageExtension.cs

```diff
@@ -0,0 +1,58 @@
+﻿using System;
+using System.Net.Http;
+using System.Text;
+using Microsoft.IdentityModel.Protocols;
+
+namespace Owin.Security.OpenIdConnect.Server.Tests {
+    public static class OpenIdConnectMessageExtension {
+        public static HttpRequestMessage ToHttpRequestMessage(this OpenIdConnectMessage openIdRequestMessage) {
+            if (openIdRequestMessage == null) {
+                throw new ArgumentNullException("openIdRequestMessage");
+            }
+
+            switch (openIdRequestMessage.RequestType) {
+                case OpenIdConnectRequestType.AuthenticationRequest:
+                    return GenerateAuthenticationRequest(openIdRequestMessage);
+                case OpenIdConnectRequestType.LogoutRequest:
+                    return GenerateLogoutRequest(openIdRequestMessage);
+                case OpenIdConnectRequestType.TokenRequest:
+                    return GenerateTokenRequest(openIdRequestMessage);
+                default:
+                    throw new ArgumentOutOfRangeException();
+            }
+        }
+
+        private static HttpRequestMessage GenerateTokenRequest(OpenIdConnectMessage openIdRequest) {
+            if (string.IsNullOrWhiteSpace(openIdRequest.TokenEndpoint)) {
+                throw new InvalidOperationException("Token endpoint should have a value when the OpenIdConnectMessage.RequestType is OpenIdConnectRequestType.TokenRequest");
+            }
+
+            var builder = !string.IsNullOrEmpty(openIdRequest.IssuerAddress) ?
```

Comment:
You shouldn't have to manually recreate an absolute URL, `HttpClient` should handle relative. Simply use `TokenEndpoint`.